tools/logparser2.py

```diff
@@ -11,86 +11,144 @@
+        meta_filename = [name for name in self.zip.namelist() if name.startswith('meta_')]
+        assert len(meta_filename) == 1
+        self.metaname = meta_filename[0]
+        self.streams = dict(_meta(self))
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.zip.close()
+
+    def open(self, filename):
+        assert '/' not in filename
+        return self.zip.open(filename)
+
+    def read(self, filename):
+        assert '/' not in filename
+        return self.zip.read(filename)
+
+
+def Log(path):
+    if path.endswith('.zip'):
+        return ZipLog(path)
+    else:
+        filename = os.path.basename(path)
+        assert filename.startswith('meta_')
+        return DirLog(path)
+
+
+def sensor_gen(path, *selected):
     """
       Yield timestamp ordered data from all selected sensors.
-      Use None for all available sensors
+      By default return data from all sensors.
     """
-    meta_dir = os.path.dirname(meta_filename)
-    filenames = {}
-    with open(meta_filename, 'rb') as meta:
-        for line in filter(lambda a: ':' in a, meta): # skip unknown lines
-            sensor, filepath = line.strip().split(':')
-            filename = os.path.basename(filepath)
-            filenames[sensor] = os.path.join(meta_dir, filename)
-
-    if selected is not None:
-        # delete filenames not in selected
-        for key in filenames.viewkeys() ^ selected:
-            del filenames[key]
-
-    # zip multiple generators
-    data = []
-    for sensor, filename in filenames.iteritems():
-        with open(filename, 'rb') as stream:
-            data.extend(timestamps_gen(sensor, stream))
-
-    data.sort()
-    for a in data:
-        yield a
-
-
-def sensor_gen_zip(filepath, selected=None):
-    with ZipFile(filepath) as zip:
-        filenames = zip.namelist()
-        meta_filename = [name for name in filenames if name.startswith('meta_')][0]
-
-        filenames = {}
-        with zip.open(meta_filename) as meta:
-            for line in filter(lambda a: ':' in a, meta): # skip unknown lines
-                sensor, filepath = line.strip().split(':')
-                filenames[sensor] = os.path.basename(filepath)
-
-        if selected is not None:
-            # delete filenames not in selected
-            for key in filenames.viewkeys() ^ selected:
-                del filenames[key]
-
-        # zip multiple generators
-        data = []
-        for sensor, filename in filenames.iteritems():
-            with zip.open(filename) as stream:
-                data.extend(timestamps_gen(sensor, stream))
-
-        data.sort()
-        for a in data:
-            yield a
+    selected = set(selected)
+    with Log(path) as log:
+        streams = []
+        for sensor in log.streams:
+            if sensor != 'can' and (len(selected) == 0 or sensor in selected):
+                gen = _stream(log, sensor)
+                streams.append(gen)
+
+        for a in _merge(streams):
```

Comment:
I have a bit of problem if to look for function _before_ the code or _after_ ... here `_stream` is before while `_merge` is after. I kind of prefer the C-style, where used functions are already defined before. (???)

Reply:
Well :smile: , I really prefer to write (and read) code top-down. The C-style is a bottom-up. The reason why it is mixed in this file is because you original started it bottom-up and I continued top-down (`_stream` is original `timestamp_gen`). I didn't want to move the code around because it looses _blame-ability_.

Let me know what you want me to do:
* top-down whole file (my preference)
* bottom-up whole file
* keep it like it is

Reply:
`top-down whole file` ... thanks